Fix bugs, improve thread safety, and enhance CI/CD

Unnatural-Dev opened the pull request:
## Summary

Comprehensive code quality improvements, bug fixes, and new features.

### Bug Fixes
- **UUID Comparison Bug** (`HoverReceiver.java:185`): Fixed `==` to `.equals()` for correct player priority detection
- **Vec3i Comparison Bug** (`HoverReceiver.java:253`): Fixed `!=` to `.equals()` preventing duplicate entries
- **Integer Overflow** (`Shop.java:555`): Added overflow protection in `buyItem()` for large transactions

### Code Quality Improvements
- **Custom Exception**: Created `InvalidConfigException` for clearer config validation errors
- **Null Safety**: Fixed `Configs.java` - proper `@Nullable` annotations with validation in getters
- **Safe UI Access**: Use `getActiveCanvas()` accessor in `Shop.java` to prevent NPE
- **Logging**: Replaced all `printStackTrace()` with `LOGGER.error()` across codebase
- **Logging**: Replaced `System.out.println` with `LOGGER.info` in `FontWidthGenerator`
- Removed debug statements and fixed TODO comments

### New Features
- **Item Blacklist**: Server admins can configure items that cannot be sold via `item_blacklist` config
- **External Link Icon**: Added visual icon to inventory view button in shop inspection UI
- **Unit Test Suite**: 51 JUnit 5 tests for data classes (`PlayerStash`, `PlayerShopBalance`, `StashEntry`)

### Thread Safety Improvements
- Replaced `HashMap` with `ConcurrentHashMap` for shared data structures
- Replaced `ArrayList` with `CopyOnWriteArrayList` for scheduled save lists

### Performance Improvements
- ArrayList capacity hint in `HoverReceiver.java` for ray casting

### CI/CD Enhancements
- Automatic dependency building for FrameworkConfig and FrameworkLib
- Gradle and Fabric Loom caching for faster builds
- Concurrency control and 30-minute timeout
- 30-day artifact retention

## Test Plan
- [x] Project builds successfully with `./gradlew build`
- [x] All 51 unit tests pass with `./gradlew test`
- [ ] Test shop creation and deletion
- [ ] Test buy/sell transactions with various items
- [ ] Test item blacklist functionality
- [ ] Test concurrent player interactions

Unnatural-Dev committed:
Fix bugs, improve thread safety, and enhance CI/CD

## Bug Fixes
- Fix UUID comparison using == instead of .equals() in HoverReceiver.java
- Fix Vec3i comparison using != instead of .equals() in HoverReceiver.java
- Add integer overflow protection in Shop.buyItem() price calculation

## Code Quality
- Remove debug System.out.println() in StashHud_MaterialList.java
- Fix TODO debug message in CommandManager.java stats command
- Replace all printStackTrace() calls with proper LOGGER.error() logging

## Thread Safety
- Replace HashMap with ConcurrentHashMap in ShopManager, StashManager, BalanceManager
- Replace ArrayList with CopyOnWriteArrayList for scheduled save lists
- Improves stability in multi-threaded Minecraft server environment

## CI/CD Improvements
- Add Gradle dependency caching for faster builds
- Add Fabric Loom cache
- Add concurrency control to cancel duplicate builds
- Add 30-minute build timeout
- Improve artifact naming with commit SHA
- Add shallow clone for faster checkout

Unnatural-Dev committed:
Add item blacklist feature and code documentation

- Add item_blacklist config field to ShopConfig.java
- Implement blacklist validation in EditUi_ItemSelector.java
- Add clarifying comments to ShopCanvas permission checks
- Document rotation calculation in EditUi_RotateButton.java
- Update MAX_PRICE documentation in Shop.java
- Fix config duplication comment in PerformanceConfig.java
- Add ArrayList capacity hint optimization in HoverReceiver.java

scmcgowen commented:
Thank you for the PR, we will review it soon

scmcgowen commented:
Also, we typically do development discussion on our discord, if you're not there, and want the link let me know. Discord discussion is not necessary to do a PR, but it may help us know what is going on.

Unnatural-Dev committed:
Add code quality improvements and unit tests

Code Quality:
- Create InvalidConfigException for better config validation errors
- Fix null safety in Configs.java (proper @Nullable annotations + validation)
- Fix logging: replace all printStackTrace() with LOGGER.error()
- Fix logging: replace System.out.println with LOGGER.info in FontWidthGenerator
- Use safe getActiveCanvas() accessor in Shop.java

New Features:
- Add external link icon design to InspectUi_InventoryViewButton
- Add unit test suite with 51 tests for data classes (JUnit 5)

Build:
- Add JUnit 5 and Mockito dependencies to build.gradle

scmcgowen reviewed (approved):
LGTM

UwUSnek commented:
- The entire thing is AI generated. We don't do that here.
- Some changes are pointless or straight up wrong  and make arbitrary decisions, removing //todos and other important comments.
- The "button design" of InspectUi_InventoryViewButton is pointless, as the element is placed behind the 3d item display. It having no design is intentional. This clearly shows that the changes have never been tested in-game by anyone.

Aside from the nonsensical ai stuff, this still has many issues.
- We are actively working on the mod, so massive pull requests that change everything are not ideal.
- The pull request is for an old version of the repository. Some of the stuff has been moved to Framework Lib and some of the issues have already been addresses properly.
- Test files are in the same directories as the actual classes. This makes the file structure even more of a mess than it already is.


<img width="1000" alt="image_2025-11-25_15-18-47" src="https://github.com/user-attachments/assets/a5084590-2a08-4891-b7e7-671f3400da08" />
<img width="512" alt="image" src="https://github.com/user-attachments/assets/a04b5ff2-e30d-4a7d-bad6-f55d5bb052ce" />

UwUSnek commented:
We appreciate contributions and are grateful to people who help make the mod better. 
That said, this is not the way to do it, and it also doesn't improve the mod.

If you have genuine, more targeted changes that are actually useful, feel free to open more pull requests.
